File: Algorithm_practice/sorting/merge_sort.cpp
// 최악의 경우에서 O(N*logN) 을 보장함
// 정확하게 반절로 가르므로..!
#include<stdio.h>

int arr[8] = {7, 6, 5, 8, 3, 9, 1, 4};
int num = 8;
int sorted[8];

void merge(int a[], int m, int middle, int n) {
    int i = m;
    int j = middle+1;
    int k = m;
    // 작은 순서대로 배열에 삽입
    while( i <= middle && j <= n) {
        if ( a[i] <= a[j] ) {
            sorted[k] = a[i];
            i++;
        } else {
            sorted[k] = a[j];
            j++;
        }
        k++;
    }
    if ( i > middle ) {
        for(int t=j; t<=n; t++) {
            sorted[k] = a[t];
            k++;
        }
    }
    else {
            for(int t=i; t<=middle; t++) {
                sorted[k] = a[t];
                k++;
            }
        }
        for(int t= m; t<=n; t++) {
            a[t] = sorted[t];
        }
}

void mergeSort(int a[], int m, int n) {
    if (m < n) { // 요소 리스트가 1보다 큰 경우
        int middle = (m+n)/2;
        mergeSort(a, m, middle);
        mergeSort(a, middle+1, n);
        merge(a, m, middle, n);
    }
}

int main(void) {
    mergeSort(arr, 0, num-1);
    for(int i=0; i< num; i++) {
        printf("%d ", arr[i]);
    }
    return 0;
}

// 기존의 데이터를 담을 추가적인 배열 공간이 필요하단 점에서 메모리 활용이 비효율적이나 속도를 보장하므로 효율적이다.

File: Algorithm_practice/Algorithm_py/merge_sort.py
arr = [3, 4, 2, 5, 1]

def merge(data, init, middle, finish):
    sorted = []
    i = init
    j = middle+1
    while i <= middle and  j <= finish: # 절반 기준으로 i나 j가 절반이나 끝을 넘어가지 않는다면
        if data[i] <= data[j]: # 만약 i번째 데이터가 j번째 데이터보다 작으면 i를 먼저 배열에 넣기
            sorted.append(data[i])
            i += 1
        else:
            sorted.append(data[j]) # j 데이터가 i데이터 보다 작으면 j데이터를 넣기
            j += 1 # 현재 j 는 넣고, 같은 i에게 현재 j를 비교!
    
    sorted.extend(data[i:middle+1])
    sorted.extend(data[j:finish+1])
    # 위 while 조건문에서 끝나는 이유가 i= middle+1 이므로
    # 위처럼 슬라이싱하면 둘 중 하나는 아무것도 추가되지 않는다.
    
    for idx, val in enumerate(sorted):
        data[init + idx] = val

def mergeSort(data, init, finish):
    if init < finish:
        middle = (init+finish)//2
        mergeSort(data, init, middle)
        mergeSort(data, middle+1, finish)
        merge(data, init, middle, finish)
    
mergeSort(arr, 0, len(arr)-1)
print(arr)

